File: test-suite/src/test/java/io/micronaut/kafka/docs/seek/aware/ConsumerSeekAwareTest.java
package io.micronaut.kafka.docs.seek.aware;

import io.micronaut.context.annotation.Property;
import io.micronaut.kafka.docs.Products;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import io.micronaut.testcontainers.kafka.Kafka;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@MicronautTest
@Property(name = "spec.name", value = "ConsumerSeekAwareTest")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ConsumerSeekAwareTest implements TestPropertyProvider {

    @Override
    public Map<String, String> getProperties() {
        return Kafka.getProperties();
    }

    @Test
    void testProductListener(ProductListener consumer) {
        await().atMost(10, SECONDS).until(() ->
            !consumer.processed.contains(Products.PRODUCT_0) &&
            consumer.processed.contains(Products.PRODUCT_1)
        );
    }
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/seek/aware/ConsumerSeekAwareTest.py
from time import sleep
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.kafka.docs.Products import Products
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductListener import ProductListener


@MicronautTest(environments=["kafka"])
@Property(name="spec.name", value="ConsumerSeekAwareTest")
class ConsumerSeekAwareTest:
    consumer: Annotated[ProductListener, Inject]

    @Test
    def test_product_listener(self):
        for _ in range(100):
            if self.seeked():
                break
            sleep(0.1)
        assert self.seeked()

    def seeked(self) -> bool:
        processed = [product.name for product in self.consumer.processed]
        return (Products.PRODUCT_0.name not in processed
                and Products.PRODUCT_1.name in processed)
